File: LeetcodeProblems/blind-75/array/maximum-product-subarray/solution.java
// Maximum Product Subarray - LeetCode #152

class Solution {
    public int maxProduct(int[] nums) {
        if (nums.length == 0) return 0;
        
        int max = nums[0];
        int min = nums[0];
        int result = nums[0];
        
        for (int i = 1; i < nums.length; i++) {
            int tempMax = Math.max(nums[i], Math.max(max * nums[i], min * nums[i]));
            min = Math.min(nums[i], Math.min(max * nums[i], min * nums[i]));
            max = tempMax;
            
            result = Math.max(result, max);
        }
        
        return result;
    }
}

File: LeetcodeProblems/blind-75/array/maximum-product-subarray/solution.py
# Maximum Product Subarray - LeetCode #152

from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        max_prod = nums[0]
        min_prod = nums[0]
        result = nums[0]
        
        for i in range(1, len(nums)):
            temp_max = max(nums[i], max_prod * nums[i], min_prod * nums[i])
            min_prod = min(nums[i], max_prod * nums[i], min_prod * nums[i])
            max_prod = temp_max
            
            result = max(result, max_prod)
        
        return result
